Fsa22 v2 73 gilded rose improve code quality part one v2 incremental abstracting

tisidro opened the pull request:
## Spec

See Issues: [FSA22V2-73](https://sparkbox.atlassian.net/browse/FSA22V2-73?atlOrigin=eyJpIjoiZjc5ODRkNDUyMDZjNGRjYWI2ZmFkZDY4ZmUyMDg2ZDkiLCJwIjoiaiJ9)
and [FSA22V2-74](https://sparkbox.atlassian.net/browse/FSA22V2-74?atlOrigin=eyJpIjoiNTBmNjM0YmZiN2M0NGZlNWIzMGFiZDcxNDY2YTM3ZTUiLCJwIjoiaiJ9)

## Description
Refactor to improve code quality (Jira Ticket 1). 

Focus on improving similar (not identical) code blocks, removing deeply nested if statements. I also want to improve code flow and readability. 

My strategy is to pull out variables where possible to reduce duplication and make it more readable. Then I am creating functions that will extract out some of the logic and remove some of the deep nesting.

Refactor to improve code quality (Jira Ticket 2). 

- No cognitive complexity issues should be present
- Code for function updateQuality has under 45 lines of code
- no similar blocks of code found that are not tests.

## Spec

Included Code Climate Errors (to Address for this PR)--
-  Avoid “deeply nested control flow statements”
- “Avoid Identical blocks of code…”

Excluded Code Climate Errors (do not Address for this PR)--
- Avoid “Similar blocks of code found…”
- Avoid “Function updateQuality has a Cognitive Complexity….”
- Avoid “Function updateQuality has 45 lines of code…”

## Validation:
- [ ] This PR has code changes, and our linters still pass.
- [ ] This PR has new code, so new tests were added or updated, and they pass.

### To Validate

- [ ] Make sure all PR Checks have passed.
- [ ] Check commit history.
- [ ] CodeClimate no longer identifies the issues enumerated in the “Included Code Climate Errors” sections listed above.

Thanks so much for your help with this PR! :)

tisidro committed:
test: add tests

add tests for increasing quality of aged Brie/aged Brie cannot exceed 50, quality of non-Sulfuras item decreases by 1, quality of item is never negative, sulfuras quality is constant, non-sulfuras items will degrade by 1.

FSA22V2-73

tisidro committed:
refactor: create item variables

create item variables to simplify & improve readability, decrease repetition.

FSA22V2-73

tisidro committed:
refactor: improve expiration conditional block

improve sell_in<0 if statement at the bottom of the code as it was a long block I understood and was a good starting point. Change expiration logic to remove confusing negative conditions, add fast return for backstage pass item quality, and separate return nothing for constantQuality variable to clarify its quality does not change.

FSA22V2-73

tisidro committed:
refactor: create quality helper functions

write functions for increasing/decreasing quality - qualityIncreaser and qualityDecreaser

FSA22V2-73

tisidro committed:
refactor: combine quality increaser/decreaser

create qualityUpdater function by adding a change_Amount input to the generic function, this way a positive or negative number can be input and the function will increment or decrement accordingly.

FSA22V2-73

jordanleven commented on src/inventory/gilded_rose.spec.js:
```diff
@@ -1,11 +1,104 @@
+  });
+
+  // quality increases by 2 when there are 10 days or less
+  it('increase quality of backstage passes with decreased sell_in', () => {
```
You can disregard this comment. Having duplicated code in tests is less of a concern because there is such an emphasis on making sure that tests are clear, straightforward, and only fail if there are failures in the source code.

jordanleven commented on src/inventory/gilded_rose.js:
```diff
@@ -22,49 +22,80 @@ const items = [
-        if (items[i].name != 'Sulfuras, Hand of Ragnaros') {
-          items[i].quality = items[i].quality - 1
+
+   //------VARIABLES-----//
```
I'm curious to hear the rationale behind this `//------VARIABLES-----//` doc. What's the reason for it?

tisidro replied:
It helps my eye locate the variables I added here. If it's clunky/silly I can remove it! :)

jordanleven replied:
It's fine to keep, but it's not a common practice since we can see variables by let, const, and var declarations. It's also might be potentially confusing if variable declarations are done outside of this block because it implies that all variables are near this annotation. If it helps you in this individual project, that's totally fine but I would recommend training your eyes to also spot variables using the variable declarations since annotations the location of variables like this is extremely uncommon in codebases.

tisidro replied:
I removed it, I see what you mean and want to nip it in the bud.

jordanleven commented on src/inventory/gilded_rose.js:
```diff
@@ -22,49 +22,80 @@ const items = [
+    const item = items[i]
+    const agedBrie = ['Aged Brie'];
+    const backstagePass = ['Backstage passes to a TAFKAL80ETC concert'];
+    const constantQuality = ['Sulfuras, Hand of Ragnaros'];
```
There are a couple of divergent variable naming conventions here. One is defining the name of the variable by a value (`agedBrie = ['Aged Brie'];`) and the other is defining the name of the variable by what it represents (`const constantQuality = ['Sulfuras, Hand of Ragnaros'];`). Naming a variable by its value is a bad idea because it kind of defeats the rationale behind variables to begin with. It's like naming a variable `cont four = 4`. Instead, I'd name the variables based on what they _represent_ (like `constantQuality`).

tisidro replied:
You're right and I'll address this! :)

tisidro committed:
refactor: implement quality updater function

add qualityUpdater function throughout code lines 1-49 to replace repetitive and redundant code specifying conditions.

FSA22V2-73

jordanleven commented on src/inventory/gilded_rose.js:
```diff
@@ -22,49 +22,80 @@ const items = [
+   
+    //create item variables to simplify & improve readability, decrease repetition, 
+    const item = items[i]
+    const agedBrie = ['Aged Brie'];
```
If these variables are immutable and static (meaning they don't change) I'd consider moving them outside of this function and at the top of this file for easy editing!

tisidro replied:
Okay I will move them. I think they were breaking and it goes to what we were discussing yesterday so I still have a bit of work to do on it.

jordanleven replied:
Just to be clear, mutable variables (i.e.,` const item = items[i]`) should stay where it is since their value requires being placed in the for loop. However, moving `agedBrie`, `backstagePass`, and `constantQuality` outside of this function would not break the application. If it does, let me know because that might be highlighting a separate issue here.

tisidro replied:
Yes, I learned that because when I moved const item it did break. I can see why as that is tied to the functionality in the for loop!

tisidro replied:
The named variables work great in global scope :)

jordanleven commented on src/inventory/gilded_rose.js:
```diff
@@ -22,49 +22,80 @@ const items = [
-          if (items[i].sell_in < 11) {
-            if (items[i].quality < 50) {
-              items[i].quality = items[i].quality + 1
+      if (item.quality < 50) {
```
🏆 Great use of a single variable here!

jordanleven commented on src/inventory/gilded_rose.js:
```diff
@@ -22,49 +22,80 @@ const items = [
+
+    //where items expire
+    //combine these into quality changer--takes in a + or - number and you can get BOTH increase or decrease in quality
+    qualityUpdater = (item, change_Amount) =>{
```
This should include a declaration.
```suggestion
    const qualityUpdater = (item, change_Amount) =>{
```

jordanleven replied:
I'd make sure you're using consistent naming conventions for variables and params. Previously, you were using camelcase (`myGreatVariable`) but here you're something similar to snake case (`my_great_variable`).

jordanleven commented on src/inventory/gilded_rose.js:
```diff
@@ -22,49 +22,80 @@ const items = [
+    
+    if (item.sell_in < 0) {
+
+      if (agedBrie.includes(items[i].name)) {
```
Oops! I think you forgot to use your new variable!
```suggestion
      if (agedBrie.includes(item.name)) {
```

jordanleven commented on src/inventory/gilded_rose.js:
```diff
@@ -22,49 +22,77 @@ const items = [
+    qualityUpdater = (item, change_Amount) =>{
+       if (item.quality < 50 && item.quality >0){
+      item.quality = item.quality + change_Amount
+    }
```
Don't forget to properly format your code! Like we chatted about yesterday, this makes it difficult to define the boundaries of a function!

tisidro replied:
I am still struggling with this. I would like to find a tutorial on formatting and practice getting better at this so it is second nature. I'm putting that on my to do list and lattice goals. Thank you!

jordanleven commented on src/inventory/gilded_rose.js:
```diff
@@ -22,49 +22,77 @@ const items = [
 */
 export function updateQuality(items) {
+    //combine these into quality changer--takes in a + or - number and you can get BOTH increase or decrease in quality
+    qualityUpdater = (item, change_Amount) =>{
```
I think you're missing a declaration here!
```suggestion
    const qualityUpdater = (item, change_Amount) =>{
```

jordanleven replied:
It looks like you might be using a combination of camelcase (`myGreatVariable`) and snake case (`my_great_variable`). I'd use camelcase for consistency.

jordanleven commented on src/inventory/gilded_rose.js:
```diff
@@ -22,49 +22,77 @@ const items = [
 */
 export function updateQuality(items) {
+    //combine these into quality changer--takes in a + or - number and you can get BOTH increase or decrease in quality
+    qualityUpdater = (item, change_Amount) =>{
```
Just a heads up that this function is still scoped inside of `updateQuality`. Yesterday, we chatted about bringing this function outside of `updateQuality` to eliminate unintended sideeffects.

theogainey replied:
@jordanleven I just chatted with @tisidro and after looking at this comment I may have l misunderstood what you were telling her the problem was. Is the problem just how this function is scoped or is it side effects in general? @tisidro  If the problem is just how the function is scoped you can disregard what I said about side effects at the end of our pairing. What you showed me fixed that problem.

jordanleven commented on src/inventory/gilded_rose.js:
```diff
@@ -22,49 +22,77 @@ const items = [
 export function updateQuality(items) {
+    //combine these into quality changer--takes in a + or - number and you can get BOTH increase or decrease in quality
+    qualityUpdater = (item, change_Amount) =>{
+       if (item.quality < 50 && item.quality >0){
```
It appears there is some hidden logic in `qualityUpdater` where the quality is only updated if the quality is less and fifty and greater than 0. Is this a strategy for a specific item or for all items?

tisidro replied:
For all items, I have placed this in a different function after some extensive refactor.

jordanleven reviewed (commented):
Hey @tisidro  — looking good! I left some comments throughout the PR. I think some of the main issues we chatted about yesterday are still in here. Take a look and let me know what you think!

tisidro committed:
fix: improve variables and fix code typo/naming

move immutable variables outside function, rename item variables to be more representative of the function, remove unnecessary comment to decrease repetitive/obvious labeling, add declaration to qualityUpdater function, rename  all variables in camelCase for consistency.

FSA22V2-73

tisidro committed:
feat: add test for quality after sell_in<0

create test that doubles quality for sell_in<0. Consolidate  tests for standard items in same code block.

FSA22V2-73

tisidro committed:
feat: helper function for Aged Brie

add helper function for Aged Brie and include code to double item as sell_in reaches zero for quality updater function.

FSA22V2-73

tisidro committed:
refactor: consolidate Aged Brie tests

nest Aged Brie tests in one describe block for better code organization.

FSA22V2-83

tisidro committed:
feat: create backstage pass helper function

create and test helper function with logic increasing quality by one if sell_in>10, by 2 if sell_in<10, by 3 if sell_in<5.

FSA22V2-83

tisidro committed:
feat: update quality for default item

add function that calls helper function to update quality for default items.

FSA22V2-83

changangus commented on src/inventory/gilded_rose.js:
```diff
@@ -21,50 +21,102 @@ const items = [
+
+    //update quality for Backstage Pass
+    if (!qualityIncFastAsExpiring.includes(item.name)) {
+      if (!constantQuality.includes(item.name)) {
```
Have you considered using a switch case statement to check the `item.name`? I think it would be more readable than the `if` statements that are checking if these arrays contain the `item.name`

changangus commented on src/inventory/gilded_rose.js:
```diff
@@ -21,50 +21,102 @@ const items = [
+    //where items expire
+    if (itemExpired(item)) {
+      if (!qualityIncWithAge.includes(item.name)) {
+        if (qualityIncFastAsExpiring.includes(item.name)) {
```
The prev comment also applies here

changangus reviewed (approved):
Tests and Linters pass, and code climate isn't showing any issues outside of your test files! Everything is looking good, we did talk about continuing to refactor some of the logic but that will be addressed later on. Nice job!
